### PR TITLE
use `reqwest::Url` instead of `String` as base url

### PR DESCRIPTION
**Pull request by paritosh-08**
Instead of appending `\<path>` to base URLs, use `Url::join("<path>")`.
JIRA: https://hasurahq.atlassian.net/browse/V3API-227

### CODE REVIEW

**Review by paf31 — APPROVED** [review]
Much nicer, thanks!

File: ndc-test/src/main.rs [hunk 1]
```diff
@@ -45,7 +45,7 @@ async fn main() {
             };
 
             let configuration = Configuration {
-                base_path: endpoint,
+                base_path: reqwest::Url::from_str(&endpoint).unwrap(),
```
Comment (paf31) [inline ndc-test/src/main.rs:48]: Try moving this into the `Commands` type please:

```rust
#[clap(
    takes_value = true,
    parse(try_from_str)
)]
endpoint: Url,
```

**Review by paf31 — APPROVED** [review]

**Comment by SamirTalwar** [comment]
Yo, this is a breaking change, as the URL must now end in a trailing slash if the path is non-empty.

We will need to check that this is reasonable.

**Comment by paf31** [comment]
> the URL must now end in a trailing slash if the path is non-empty.

So you get an error from `reqwest` for something like `http://domain/path` now?

**Comment by paritosh-08** [comment]
My bad; this is the expected behaviour by the library: https://github.com/servo/rust-url/issues/285.

Should we do something like the suggestion in the comment: https://github.com/servo/rust-url/issues/285#issuecomment-285346010 (add trailing slash if not present)?

**Comment by paritosh-08** [comment]
> > the URL must now end in a trailing slash if the path is non-empty.
> 
> So you get an error from `reqwest` for something like `http://domain/path` now?

No, while joining `query` to the above url, the `URL::join()` will return something like `http://domain/query`. However, for `http://domain/path/`, the joining will return `http://domain/path/query`.

**Comment by paf31** [comment]
And what if you join with `/query`?

**Comment by paritosh-08** [comment]
> And what if you join with `/query`? Same issue?

Yes.

**Pull request merged**